### ISSUE TITLE
MySQL ACL not working 

### ISSUE THREAD

**Issue by inteldon** [emqx/emqx#7072]
Using emqx version 4.3.12 
MySQL version 8.0.28 

Follow instructions on [this link](https://www.emqx.io/docs/en/v4.3/advanced/acl-mysql.html#default-table-structure). 

Created two users Alice/Bob using this command: 
```
INSERT INTO `mqtt_user` ( `username`, `password`, `salt`, `is_superuser`)
VALUES
	('alice', 'efa1f375d76194fa51a3556a97e641e61685f914d446979da50a551a4333ffd7', NULL, 0);
```

```
INSERT INTO `mqtt_user` ( `username`, `password`, `salt`, `is_superuser`)
VALUES
	('bob', 'efa1f375d76194fa51a3556a97e641e61685f914d446979da50a551a4333ffd7', NULL, 0);
```

Created an ACL rule to allow users to publish to their own topic "data/%u" 
`INSERT INTO mqtt_acl (allow, ipaddr, username, clientid, access, topic) VALUES (1, NULL, '$all', NULL, 1, '/data/%u');`

Expected: Alice can only publish/subscribe to "data/alice" and Bob can only publish/subscribe to "data/bob"
Result: Both Alice/Bob can publish/subscribe to "data/alice" and "data/bob"

I also tried "access" level 3, to allow Alice/Bob publish and subscribe permissions, but the result was the same.

**Comment by inteldon** [comment]
I tried using compound rules for ACL: 
1. Allow Alice to publish/subscribe to "data/alice/#" 
2. Deny everyone else to publish/subscribe to "data/alice/#" 

```
-- Allow Alice to pub/sub to her topics
INSERT INTO mqtt_acl (allow, ipaddr, username, clientid, access, topic) VALUES (1, NULL, 'alice', NULL, 3, '/data/alice/#');

-- Deny everyone else to pub/sub
INSERT INTO mqtt_acl (allow, ipaddr, username, clientid, access, topic) VALUES (0, NULL, '$all', NULL, 3, '/data/alice/#');
```
Result: I logged in as a different user, and I could still publish/subscribe to Alice's topics. 
Expected: Only Alice can publish/subscribe to her topic "/data/alice/#"

**Comment by ieQu1** [comment]
Hello,

Can you show `acl.conf`? By default it contains permissive rules.
What is the value of `acl_nomatch` parameter in `emqx.conf`? (It should be `deny` to get the expected results)

**Comment by inteldon** [comment]
Hi, 

Thanks. I shared my files here in this message for your reference. 
[emqx_auth_mysql.txt](https://github.com/emqx/emqx/files/8122432/emqx_auth_mysql.txt)
[acl.txt](https://github.com/emqx/emqx/files/8122433/acl.txt)
[emqx.txt](https://github.com/emqx/emqx/files/8122434/emqx.txt)

Please note the following: 

1. Even though some features are enabled in acl.conf, such as deny access to "SYS" topic, I was still able to subscribe to the "SYS" topic with the "alice" username. This should not be allowed by acl.conf. 

![image](https://user-images.githubusercontent.com/28794155/155274326-0aa38994-08c1-4d85-b360-9c2d91536b7e.png)

2. I disabled the internal acl through the dashboard (Modules). 

![image](https://user-images.githubusercontent.com/28794155/155274668-0f8967d8-359e-4d21-ad52-46d5a48430d1.png)


3. I verified that my MySQL plugin is running through the dashboard 

![image](https://user-images.githubusercontent.com/28794155/155274610-e133b9a2-5e4d-41f7-b434-15f029ff0316.png)

**Comment by ieQu1** [comment]
I see that you have `acl_nomatch = allow` in emqx.config file, you probably want to change it to `deny`.
Also in `acl.conf` you have `{allow, all}.` rule. This needs to be removed, because likely it overrides the previous rules.
Finally, can you show the output of `emqx_ctl plugins list` command?

**Comment by inteldon** [comment]
Hi Thanks for looking into the files. This is running in docker containers. I just noticed that each time I change the acl file and commit my changes the acl.conf file and the emqx.conf file go back to default mode automatically. I tried running the reload command but I get errors. 

`./bin/emqx_ctl acl reload 
`
```
ls: /opt/emqx/data/configs/vm.*.args: No such file or directory
No vm.*.args config file found in /opt/emqx/data/configs/
```

**Comment by zmstone** [comment]
hi @inteldon 
that's because `/opt/emqx/data` is a docker VOLUME
docker does not commit data in volumes.

what you can do is to add your file to a different place
and use `docker run -e EMQX_ACL_FILE=/path/to/a/different/place` to override the path.

Please note that `/opt/emqx/etc` is also a volume
but this is being fixed in the next patch release v4.3.13 and v4.4.2

**Comment by inteldon** [comment]
Thanks and good point. I'll give this a try. Should I do the same for the emqx.conf file? What environment variables should I use for that?

**Comment by zmstone** [comment]
for config file, if you need to commit changes, the only way (before the next release) is to copy the etc dir to another place and override withe `RUNNER_ETC_DIR`
Like this: https://github.com/emqx/emqx/issues/7093#issuecomment-1047490799

**Comment by inteldon** [comment]
Hi and thanks for clarifying. I tried to make a new directory in the container and use those options, although the change persists now after the commit, the container doesn't run and exits as soon as I start it. 

Thanks for mentioning the new changes that are coming to the container. I'll give that a try when the changes are released. For now I reverted back to non-container based broker and it works fine.
